recorderproxy/recordcontext.go

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright 2019 National Library of Norway.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package recorderproxy
+
+import (
+	"context"
+	"github.com/nlnwa/veidemann-api-go/browsercontroller/v1"
+	"github.com/nlnwa/veidemann-api-go/commons/v1"
+	"github.com/nlnwa/veidemann-api-go/config/v1"
+	"github.com/nlnwa/veidemann-api-go/contentwriter/v1"
+	"github.com/nlnwa/veidemann-api-go/frontier/v1"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"io"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+	"sync/atomic"
+	"time"
+)
+
+type recordContext struct {
+	// Will contain the client request from the proxy
+	Req *http.Request
+	// Will contain the remote server's response (if available. nil if the request wasn't send yet)
+	Resp *http.Response
+	//RoundTripper RoundTripper
+	// will contain the recent error that occurred while trying to send receive or parse traffic
+	Error error
+	// Will connect a request to a response
+	session int64
+	proxy   *RecorderProxy
+
+	conn              Connections
+	cwcCancelFunc     context.CancelFunc
+	cwc               contentwriter.ContentWriter_WriteClient
+	bcc               browsercontroller.BrowserController_DoClient
+	bccCancelFunc     context.CancelFunc
+	bccMsgChan        chan *browsercontroller.DoReply
+	uri               *url.URL
+	IP                string
+	FetchTimesTamp    time.Time
+	collectionRef     *config.ConfigRef
+	meta              *contentwriter.WriteRequest_Meta
+	crawlLog          *frontier.CrawlLog
+	replacementScript *config.BrowserScript
+	closed            bool
+	foundInCache      bool
+	precludedByRobots bool
+}
+
+var i int
+
+func NewRecordContext(proxy *RecorderProxy) *recordContext {
+	ctx := &recordContext{
+		session: atomic.AddInt64(&proxy.sess, 1),
+		proxy:   proxy,
+		conn:    proxy.conn,
+	}
+
+	return ctx
+}
+
+func (ctx *recordContext) init(req *http.Request) *recordContext {
+	cwcCtx, cwcCancel := context.WithTimeout(context.Background(), ctx.proxy.ConnectionTimeout)
+	cwc, err := ctx.proxy.conn.ContentWriterClient().Write(cwcCtx)
+	if err != nil {
+		log.Fatalf("Error connecting to content writer, cause: %v", err)
```

Comment:
`log.Fatalf` is equivalent to `log.Printf` followed by `os.Exit(1)`. So code following this statement does not execute.